README.md
# Card Warp

This is a library used to detect a card on a picture and warp it in order to generate an image of a perfectly straight, horizontal image of a card, that can be processed further by an OCR.

This functionality is achieved using opencv. Two algorithms are implemented: One that uses feature matching algorithms to detect a card, and one that uses Canny edge detection combined with a Hough transform to find a card.

## Comparison

### Feature Matching

This method searches "interesting points" (features) in a reference image of a card that only contains features that remain constant over all cards.

This step is repeated for the actual image on which a card should be detected. Afterwards, an algorithm tries to find a homography between the two sets of features.

If enough features on the input image match the reference image, the borders and corner points of the card can be calculated from their arrangement.  

**Pros:**
- The card can be held in very weird positions and still be recognized
- Can detect a card anywhere in the picture
- Only a card is detected that actually matches the reference image
- Very precise detection if enough features are present

**Cons:**
- Uniform cards not containing enough features are rarely detected
- Requires reference images
- Can not detect any type of card
- Slower & consumes much more resources than Hough Transform

### Hough Transform

This method only searches in the border regions of a rectangle of a given size in the center of the picture, which makes it a lot faster than Feature Matching.

There are four rectangular regions at the four borders of the detection rectangle which the detection algorithm takes place in.

Each of these regions gets edged using Canny, and a Hough Transform applied on the result.

From the lines returned by Hough the best line is calculated, so the result is exactly one line for each region.

Four corner points are then by determined by calculating the intersections of said lines. 

**Pros:**
- Can detect any type of card out of the box
- Can detect very homogeneous cards as well
- Very fast and resource efficient

**Cons:**
- Card has to be in the detection rectangle and must not be tilted excessively
- Might in rare cases misinterpret lines on or above the card as the cards border
- No restriction for the card type possible

## Install

- `npm install @techteamer/card-warp`

## Usage

### Feature Matching

#### Reference image

The reference image is an image of the card, that should be matched, with everything, that changes over different cards, either blurred out, or removed, while everything, that stays the same, left in it.
Blurring works, because it removes sharp edges, which would otherwise be detected by the feature detection algorithm.
 
The feature detection algorithm detects edges, so if there are edges on the reference image, that might not occur on the image, that should be matched, the output will be less accurate.

For instance, on an ID card, the face, name and other personal data should be removed, while labels like the country name, or the field descriptions should be left in.

**Example:**

![Example Reference Image](test/feature-matcher/images/reference.jpg)

#### How to use in a project

Use `FeatureMatcher::generateDescriptors` to generate descriptors for a reference image that should be matched.

An image buffer together with the descriptors are then piped into `FeatureMatcher#getCard` to generate an image of a straight, horizontal card.

The result is a Promise resolving in an object with the key `card` being a buffer of the generated image as well as the key `probability` being a probability whether the result is actually a card.

The probability is the quotient of the number of found points matching the geometric model and the total number of found points. Have a look at the source code for more information.

**Example:**

```javascript
const fs = require('fs')
const FeatureMatcher = require('@techteamer/card-warp').FeatureMatcherWarper

let detector = new FeatureMatcher()
let descriptors = detector.generateDescriptors('features/id_new.jpg')

detector.getCard(fs.readFileSync('input.jpg'), descriptors)
  .then(obj => {
    if (obj === false)
      return console.log('No card found')

    console.log('Probability:', (obj.probability * 100).toFixed(2) + '%')

    fs.writeFileSync('output.png', obj.card)
  })
```

#### Demo examples

Each directory in `./test/images` contains the test images as well as an image of the expected result (`result.jpg`).

`./test/images/reference.jpg` is a reference image of the sample card that will be matched in the test images.

`docker-compose up test` tests each image, checks if it is similar enough to the expected result as well as generates a `graph-*.jpg` for each tested image.
The graph is an image that shows similarities between the expected result and the actual result, so a human can take a look at the output as well.

### Hough Transform

#### How to use in a project

Feed `HoughTransform#getCard` with an image buffer in order to generate an image of a straight, horizontal card.

The result is a Promise resolving in a buffer of the generated image.

**Example:**

```javascript
const fs = require('fs')
const HoughTransform = require('@techteamer/card-warp').HoughTransformWarper

let detector = new HoughTransform({
  "detectionRectangleWidth": 450,
  "detectionRectangleHeight": 300,
  "detectionWidth": 50
})

detector.getCard(fs.readFileSync('input.jpg'))
  .then(cardBuffer => {
    fs.writeFileSync('output.png', cardBuffer)
  })
```

#### Demo

- `npm install`
- `cd test/hough-transform`
- `node test`

The results can be found in `test/hough-transform/output`.

## API

### `FeatureMatcherWarper`

This is the only class this library exposes and it's used to generate feature descriptors as well as detect a card on an input image, warp it and output a straight image of said card.

#### `#generateDescriptors (path, downscaleWidth = 1000): Promise<Object>`

**Parameters:**
- `path: string`: The path to the image on the local filesystem
- `downscaleWidth: number`: The width images should be downscaled to if they exceed it

This function generates and returns corners, key points and features descriptors of an image.
It is separate, so this function can be run once at the application startup for every reference image, without having to re-generate the same descriptors every time an image is matched.

#### `#getCard (inputBuffer, reference, outputWidth = 500): Promise<Buffer>`

**Parameters:**
- `inputBuffer: Buffer`: The input image as a buffer
- `reference: Object`: The descriptors acquired by `FeatureMatcherWarper#generateDescriptors`
- `outputWidth: number`: The desired width of the output image

This function detects and warps an output image based on the reference descriptors and returns a buffer of the warped card as PNG.

### `HoughTransformWarper`

#### `constructor (options = {})`

**Options:**

```javascript
options = {
  detectionRectangleWidth: 320,   // Default width of the detection rectangle
  detectionRectangleHeight: 240,  // Default height of the detection rectangle
  detectionWidth: 50,             // Default width of the detection border regions

  /*
   * Do not set these options if you don't know what you're doing.
   * It might mess things up.
   */
  cannyLowerThreshold: 150,       // Lower threshold applied to Canny
  cannyThresholdRatio: 3,         // higherThreshold = cannyLowerThershold * cannyThresholdRatio

  houghRho: 1,                    // Rho applied to Hough Transform
  houghTheta: Math.PI / 180,      // Theta applied to Hough Transform
  houghThreshold: 75              // Threshold applied to Hough Transform
}
```

> Any missing key in the options will be replaced by the default options above.

#### `#getCard (options = {}): Promise<Buffer>`

 **Parameters:**
 - `inputBuffer: Buffer`: The input image as a buffer
 - *`_options: Object`: The default options passed in the constructor will be replaced by this parameter for the current call*
 - *`_options.outputWidth: number`: Desired width of the output image*
 - *`_options.outputHeight: number`: Desired height of the output image*
 - *`_options.detectionRectangleWidth: number`: Detection rectangle width*
 - *`_options.detectionRectangleHeight: number`: Detection rectangle height*
 - *`_options.detectionWidth: number`: Width of the detection border regions*

This function detects and warps an output image based on the reference descriptors and returns a buffer of the warped card as PNG.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [1.0.0] - 2019-03-11
### Added
- Card detection functionality
- Card warping functionality
- Tests

## [1.0.1] - 2019-04-01
### Changed
- .npmignore


test/test.js
const fs = require('fs')
const test = require('ava')
const path = require('path')
const cv = require('opencv4nodejs')
const CardWarp = require('../src/CardWarp')

let cardWarp = new CardWarp()

let referenceDescriptors
let resultGraphs = []

test.before('Generate Descriptors', async () => {
  referenceDescriptors = await cardWarp.generateDescriptors(path.resolve('test/images/reference.jpg'))
})

for (let name of [ 'Emerson Allen', 'Jonathan Pelchat', 'Joshua Gould' ]) {
  let samplePath = path.join(path.resolve('test/images'), name.toLowerCase().split(' ').join('-'))
  let graphs = []

  test(name, async t => {
    let { resultDescriptors, images } = await readDir(samplePath)

    for (let { name, buffer } of images) {
      let card = await cardWarp.getCard(buffer, referenceDescriptors)

      t.is(card.probability > 0.4, true, `Probability of ${name} greater than 0.4`)

      let { similarity, graph } = await checkSimilarities(card.card, resultDescriptors)

      t.is(similarity > 0.4, true, `Similarity of ${name} and result.jpg greater than 0.4`)

      cv.imwrite(path.join(samplePath, `graph-${name}`), graph)
    }
  })

  resultGraphs.push(graphs)
}

async function readDir (dir) {
  let files = fs.readdirSync(dir).filter(fileName => path.extname(fileName) === '.jpg' && !fileName.startsWith('graph-'))
  let result = {}

  if (!files.includes('result.jpg')) {
    return { resultDescriptors: false, images: false }
  }

  result.resultDescriptors = await cardWarp.generateDescriptors(path.resolve(dir, 'result.jpg'))
  result.images = files
    .filter(fileName => fileName !== 'result.jpg')
    .map(name => ({
      name,
      buffer: fs.readFileSync(path.resolve(dir, name))
    }))

  return result
}

async function checkSimilarities (inputBuffer, result) {
  let detector = new cv.SIFTDetector({ nFeatures: 4000 })
  let inputMat = await cv.imdecodeAsync(inputBuffer)

  let {
    image: resultImage,
    keyPoints: resultKeyPoints,
    descriptors: resultDescriptors
  } = result

  let inputKeyPoints = await detector.detectAsync(inputMat)
  let inputDescriptors = await detector.computeAsync(inputMat, inputKeyPoints)

  if (inputDescriptors.rows === 0) {
    return {
      similarity: 0
    }
  }

  let rawMatches = await cv.matchKnnFlannBasedAsync(resultDescriptors, inputDescriptors, 2)
  let goodMatches = []

  let resultPoints = []
  let inputPoints = []

  let graph

  let mask

  let inliers = 0

  for (let match of rawMatches) {
    if (match.length === 2 && match[0].distance < match[1].distance * 0.75) {
      goodMatches.push(match[0])

      resultPoints.push(resultKeyPoints[match[0].queryIdx].point)
      inputPoints.push(inputKeyPoints[match[0].trainIdx].point)
    }
  }

  let sortedMatches = goodMatches
    .sort((m1, m2) => m1.distance - m2.distance)
    .slice(0, 20)

  graph = cv.drawMatches(resultImage, inputMat, resultKeyPoints, inputKeyPoints, sortedMatches)

  ;({ mask } = cv.findHomography(resultPoints, inputPoints, cv.RANSAC, 4))

  for (let i = 0; i < inputPoints.length; i++) {
    if (mask.at(i, 0) === 1) {
      inliers++
    }
  }

  return {
    similarity: inliers / inputPoints.length,
    graph
  }
}


test/feature-matcher/test.js
const fs = require('fs')
const test = require('ava')
const path = require('path')
const cv = require('opencv4nodejs')
const CardWarp = require('../../src/FeatureMatcherWarper')

let cardWarp = new CardWarp()

let referenceDescriptors
let resultGraphs = []

test.before('Generate Descriptors', async () => {
  referenceDescriptors = await cardWarp.generateDescriptors(path.resolve('test/images/reference.jpg'))
})

for (let name of [ 'Emerson Allen', 'Jonathan Pelchat', 'Joshua Gould' ]) {
  let samplePath = path.join(path.resolve('test/images'), name.toLowerCase().split(' ').join('-'))
  let graphs = []

  test(name, async t => {
    let { resultDescriptors, images } = await readDir(samplePath)

    for (let { name, buffer } of images) {
      let card = await cardWarp.getCard(buffer, referenceDescriptors)

      t.is(card.probability > 0.4, true, `Probability of ${name} greater than 0.4`)

      let { similarity, graph } = await checkSimilarities(card.card, resultDescriptors)

      t.is(similarity > 0.4, true, `Similarity of ${name} and result.jpg greater than 0.4`)

      cv.imwrite(path.join(samplePath, `graph-${name}`), graph)
    }
  })

  resultGraphs.push(graphs)
}

async function readDir (dir) {
  let files = fs.readdirSync(dir).filter(fileName => path.extname(fileName) === '.jpg' && !fileName.startsWith('graph-'))
  let result = {}

  if (!files.includes('result.jpg')) {
    return { resultDescriptors: false, images: false }
  }

  result.resultDescriptors = await cardWarp.generateDescriptors(path.resolve(dir, 'result.jpg'))
  result.images = files
    .filter(fileName => fileName !== 'result.jpg')
    .map(name => ({
      name,
      buffer: fs.readFileSync(path.resolve(dir, name))
    }))

  return result
}

async function checkSimilarities (inputBuffer, result) {
  let detector = new cv.SIFTDetector({ nFeatures: 4000 })
  let inputMat = await cv.imdecodeAsync(inputBuffer)

  let {
    image: resultImage,
    keyPoints: resultKeyPoints,
    descriptors: resultDescriptors
  } = result

  let inputKeyPoints = await detector.detectAsync(inputMat)
  let inputDescriptors = await detector.computeAsync(inputMat, inputKeyPoints)

  if (inputDescriptors.rows === 0) {
    return {
      similarity: 0
    }
  }

  let rawMatches = await cv.matchKnnFlannBasedAsync(resultDescriptors, inputDescriptors, 2)
  let goodMatches = []

  let resultPoints = []
  let inputPoints = []

  let graph

  let mask

  let inliers = 0

  for (let match of rawMatches) {
    if (match.length === 2 && match[0].distance < match[1].distance * 0.75) {
      goodMatches.push(match[0])

      resultPoints.push(resultKeyPoints[match[0].queryIdx].point)
      inputPoints.push(inputKeyPoints[match[0].trainIdx].point)
    }
  }

  let sortedMatches = goodMatches
    .sort((m1, m2) => m1.distance - m2.distance)
    .slice(0, 20)

  graph = cv.drawMatches(resultImage, inputMat, resultKeyPoints, inputKeyPoints, sortedMatches)

  ;({ mask } = cv.findHomography(resultPoints, inputPoints, cv.RANSAC, 4))

  for (let i = 0; i < inputPoints.length; i++) {
    if (mask.at(i, 0) === 1) {
      inliers++
    }
  }

  return {
    similarity: inliers / inputPoints.length,
    graph
  }
}


test/hough-transform/test.js
const fs = require('fs')
const FeatureMatcherWarper = require('../../src').FeatureMatcherWarper

/* let detector = new HoughTransformWarper({
  detectionRectangleWidth: 450,
  detectionRectangleHeight: 300,
  detectionWidth: 50
})

let detector = new HoughTransformWarper({
  detectionRectangleWidth: 460,
  detectionRectangleHeight: 290,
  detectionWidth: 50
}) */

let detector = new FeatureMatcherWarper()

async function main () {
  let idBackFeatures = await detector.generateDescriptors('new_back.png')

  if (!fs.existsSync('output')) {
    fs.mkdirSync('output')
  }

  /* for (let file of fs.readdirSync('images')) {
    let input = fs.readFileSync(`images/${file}`)
    let warped = await detector.getCard(input, idBackFeatures)

    if (!warped) {
      console.log('No card found')
      continue
    }

    fs.writeFileSync(`output/${file}`, warped)
  } */

  let input = fs.readFileSync('images/new_back.png')

  let warped = await detector.getCard(input, idBackFeatures)

  console.log(warped.probability)

  fs.writeFileSync('output/output.png', warped.card)
}

main()


src/CardWarp.js
const cv = require('opencv4nodejs')

module.exports = class CardWarp {
  /**
   * Initiates the card detector.
   */
  constructor () {
    this.detector = new cv.SIFTDetector({ nFeatures: 4000 })
  }

  /**
   * Detects a card on a given image
   * @param {Buffer} inputBuffer Input image buffer
   * @param {Object} reference The descriptors of the reference image, acquired by CardWarp::generateDescriptors
   * @param {number} outputWidth Width of the output picture
   * @returns {Buffer} Returns a PNG image buffer
   */
  async getCard (inputBuffer, reference, outputWidth = 500) {
    let inputMat = await cv.imdecodeAsync(inputBuffer)

    let {
      image: referenceImage,
      corners: referenceCorners,
      keyPoints: referenceKeyPoints,
      descriptors: referenceDescriptors
    } = reference

    let inputKeyPoints = await this.detector.detectAsync(inputMat)
    let inputDescriptors = await this.detector.computeAsync(inputMat, inputKeyPoints)

    let rawMatches

    let referencePoints = []
    let inputPoints = []

    let homography
    let mask

    let cardPointsMat
    let cardPoints = []

    let inliers = 0

    let warped
    let warpColorCorrected

    if (inputDescriptors.rows === 0) {
      return false
    }

    rawMatches = await cv.matchKnnFlannBasedAsync(referenceDescriptors, inputDescriptors, 2)

    for (let match of rawMatches) {
      if (match.length === 2 && match[0].distance < match[1].distance * 0.75) {
        let m = match[0]

        referencePoints.push(referenceKeyPoints[m.queryIdx].point)
        inputPoints.push(inputKeyPoints[m.trainIdx].point)
      }
    }

    ({ homography, mask } = cv.findHomography(referencePoints, inputPoints, cv.RANSAC, 4))

    for (let i = 0; i < inputPoints.length; i++) {
      if (mask.at(i, 0) === 1) {
        inliers++
      }
    }

    if (homography.cols !== 3) {
      return false
    }

    cardPointsMat = await referenceCorners.perspectiveTransformAsync(homography)

    for (let r = 0; r < cardPointsMat.rows; r++) {
      for (let c = 0; c < cardPointsMat.cols; c++) {
        cardPoints.push(new cv.Point2(cardPointsMat.at(r, c).x, cardPointsMat.at(r, c).y))
      }
    }

    warped = await warp(inputMat, cardPoints, outputWidth, ~~((outputWidth / referenceImage.cols) * referenceImage.rows))
    warpColorCorrected = await colorCorrect(warped)

    return {
      card: cv.imencode('.png', warpColorCorrected),
      probability: inliers / inputPoints.length
    }
  }

  /**
   * Generates corners, key points and feature descriptors of an image
   * @param {String} path The path to the image on the local filesystem
   * @param {number} downscaleWidth The width images should be downscaled to if they exceed it
   * @returns {Object} An object containing the corners, key points and feature descriptors of the reference images
   */
  async generateDescriptors (path, downscaleWidth = 1000) {
    let image = await cv.imreadAsync(path)

    let corners
    let keyPoints
    let descriptors

    if (image.cols > downscaleWidth) {
      image = await image.resizeAsync(~~(downscaleWidth / image.cols * image.rows), downscaleWidth)
    }

    corners = new cv.Mat([[
      [0, 0, 1],
      [image.cols, 0, 1],
      [image.cols, image.rows, 1],
      [0, image.rows, 1]
    ]], cv.CV_32FC2)

    keyPoints = await this.detector.detectAsync(image)
    descriptors = await this.detector.computeAsync(image, keyPoints)

    return {
      image,
      corners,
      keyPoints,
      descriptors
    }
  }
}

/**
 * Color-corrects a Mat
 * @param {cv.Mat} inputMat Input Mat
 * @returns {Promise<cv.Mat>}
 */
async function colorCorrect (inputMat) {
  let gray = inputMat.cvtColor(cv.COLOR_BGR2GRAY)

  let histogram = cv.calcHist(gray, [{
    channel: 0,
    bins: 256,
    ranges: [ 0, 256 ]
  }])

  let accumulator = []

  accumulator.push(histogram.at(0, 0))

  for (let i = 1; i < 256; i++) {
    accumulator.push(accumulator[i - 1] + histogram.at(i, 0))
  }

  let max = accumulator[accumulator.length - 1]

  let minGray = 0
  while (accumulator[minGray] < 1) {
    minGray++
  }

  let maxGray = 255
  while (accumulator[maxGray] >= (max - 1)) {
    maxGray--
  }

  let inputRange = maxGray - minGray
  let alpha = 255 / inputRange
  let beta = -minGray * alpha

  let result = inputMat.convertTo(-1, alpha, beta + (Math.abs(beta) * 1.5))

  return result
}

/**
 * Warps four points on a Mat to form a rectangle
 * @param {cv.Mat} inputMat Input Mat that should be warped
 * @param {cv.Point2[]} points Points on the Mat
 * @param {number} width Width of the returned image
 * @param {number} height Height of the returned image
 * @returns {Promise<cv.Mat>}
 */
async function warp (inputMat, points, width, height) {
  let trans = cv.getPerspectiveTransform(points, [
    new cv.Point2(0, 0),
    new cv.Point2(width, 0),
    new cv.Point2(width, height),
    new cv.Point2(0, height)
  ])

  let warped = await inputMat.warpPerspectiveAsync(trans, new cv.Size(width, height))

  return warped
}


src/HoughTransformWarper.js
const cv = require('opencv4nodejs')

const DEFAULT_OPTIONS = {
  detectionRectangleWidth: 320,
  detectionRectangleHeight: 240,
  detectionWidth: 50,
  outputWidth: 500,

  cannyLowerThreshold: 50,
  cannyThresholdRatio: 3,

  houghRho: 1,
  houghTheta: Math.PI / 180,
  houghThreshold: 75
}

module.exports = class HoughTransformWarper {
  /**
   * @param {Object} [options] Default options are replaced by this
   * @param {number} [options.outputWidth] Desired width of the output image.
   * @param {number} [options.outputHeight] Desired height of the output image
   * @param {number} [options.detectionRectangleWidth] Detection rectangle width
   * @param {number} [options.detectionRectangleHeight] Detection rectangle height
   * @param {number} [options.detectionWidth] Width of the detection border regions
   */
  constructor (options = {}) {
    this.options = Object.assign({}, DEFAULT_OPTIONS, options)
  }

  /**
   * Detects a card within the detection rectangle configured in the constructor and returns a straightly-warped image of said card
   * @param {Buffer} inputBuffer Input image buffer
   * @param {Object} [_options] The default options passed in the constructor will be replaced by this parameter for the current call
   * @param {number} [_options.outputWidth] Desired width of the output image.
   * @param {number} [_options.outputHeight] Desired height of the output image
   * @param {number} [_options.detectionRectangleWidth] Detection rectangle width
   * @param {number} [_options.detectionRectangleHeight] Detection rectangle height
   * @param {number} [_options.detectionWidth] Width of the detection border regions
   * @returns {Promise<Buffer>} PNG image buffer
   */
  async getCard (inputBuffer, _options = {}) {
    let options = Object.assign({}, this.options, _options)

    if (!options.hasOwnProperty('outputHeight')) {
      options.outputHeight = Math.abs(options.outputWidth * (options.detectionRectangleHeight / options.detectionRectangleWidth))
    }

    let inputMat = await cv.imdecodeAsync(inputBuffer)
    let region = getDetectionRectangle(inputMat, options.detectionRectangleWidth, options.detectionRectangleHeight)
    let blurredRegion = await region.gaussianBlurAsync(new cv.Size(5, 5), 0)

    let borderRegions = getDetectionBorders(blurredRegion, options.detectionWidth)

    let bestLines = {}
    let cornerPoints = {}

    let warped

    for (let borderRegion of borderRegions) {
      let adaptiveThresh = await borderRegion.regionMat.cvtColor(cv.COLOR_BGR2GRAY).adaptiveThresholdAsync(255, cv.ADAPTIVE_THRESH_MEAN_C, cv.THRESH_BINARY, 3, 0)
      let blur = await adaptiveThresh.gaussianBlurAsync(new cv.Size(9, 9), 0)
      let thresh = await blur.thresholdAsync(0, 255, cv.THRESH_BINARY + cv.THRESH_OTSU)

      let tempEdge = await thresh.cannyAsync(30, 150)

      let tempLines = await getLines(tempEdge, this.options.houghRho, this.options.houghTheta, 75)

      for (let line of tempLines) {
        borderRegion.regionMat.drawLine(line.p1, line.p2, new cv.Vec3(255, 0, 0), 1)
      }

      /* cv.imwrite(`output/${borderRegion.type}.png`, borderRegion.regionMat)
      cv.imwrite(`output/${borderRegion.type}-thresh.png`, tempEdge) */

      let edge = await borderRegion.regionMat.cannyAsync(this.options.cannyLowerThreshold, this.options.cannyLowerThreshold * this.options.cannyThresholdRatio)
      let lines = await getLines(edge, this.options.houghRho, this.options.houghTheta, this.options.houghThreshold)
      let bestLine

      if (tempLines.length === 0) {
        return null
      }

      bestLine = findBestLine(tempLines)

      switch (borderRegion.type) {
        case BorderRegion.TOP:
          bestLines.top = bestLine
          break
        case BorderRegion.LEFT:
          bestLines.left = bestLine
          break
        case BorderRegion.BOTTOM:
          bestLines.bottom = new Line(
            new cv.Point2(bestLine.p1.x, bestLine.p1.y + region.rows - options.detectionWidth),
            new cv.Point2(bestLine.p2.x, bestLine.p2.y + region.rows - options.detectionWidth),
            bestLine.slope
          )
          break
        case BorderRegion.RIGHT:
          bestLines.right = new Line(
            new cv.Point2(bestLine.p1.x + region.cols - options.detectionWidth, bestLine.p1.y),
            new cv.Point2(bestLine.p2.x + region.cols - options.detectionWidth, bestLine.p2.y),
            bestLine.slope
          )
          break
      }
    }

    if (!(bestLines.top && bestLines.left && bestLines.bottom && bestLines.right)) {
      return null
    }

    cornerPoints.topLeft = findIntersection(bestLines.top, bestLines.left)
    cornerPoints.topRight = findIntersection(bestLines.top, bestLines.right)
    cornerPoints.bottomLeft = findIntersection(bestLines.bottom, bestLines.left)
    cornerPoints.bottomRight = findIntersection(bestLines.bottom, bestLines.right)

    for (let points of [ cornerPoints.topLeft, cornerPoints.topRight, cornerPoints.bottomLeft, cornerPoints.bottomRight ]) {
      if (points.x === 0 || points.y === 0) {
        return null
      }
    }

    warped = await warp(region, cornerPoints, options.outputWidth, options.outputHeight)

    return cv.imencodeAsync('.png', warped)
  }
}

/**
 * Applies Hough transform in order to find lines on a provided edged input Mat
 * @param {Mat} edgesMat Edged input Mat
 * @param {number} houghRho Rho applied to Hough Transform
 * @param {number} houghTheta Theta applied to Hough Transform
 * @param {number} houghThreshold Threshold applied to Hough Transform
 * @returns {Promise<Line[]>} Array of lines
 */
async function getLines (edgesMat, houghRho, houghTheta, houghThreshold) {
  let hough = await edgesMat.houghLinesAsync(houghRho, houghTheta, houghThreshold)

  return hough.map(line => {
    let rho = line.x
    let theta = line.y
    let a = Math.cos(theta)
    let b = Math.sin(theta)
    let x0 = a * rho
    let y0 = b * rho

    let p1 = new cv.Point2(Math.round(x0 + 1000 * (-b)), Math.round(y0 + 1000 * a))
    let p2 = new cv.Point2(Math.round(x0 - 1000 * (-b)), Math.round(y0 - 1000 * a))

    return new Line(p1, p2)
  })
}

/**
 * Finds the two lines in the array with the smallest slope difference and returns their average
 * @param {Line[]} lines
 */
function findBestLine (lines) {
  /*
  let minSlopeDiff
  let minDiffLines

  for (let line1 of lines) {
    for (let line2 of lines) {
      let slopeDiff = Math.abs(line1.slope - line2.slope)

      //  if (slopeDiff === 0) {
      //   continue
      // }

      if (!minSlopeDiff || slopeDiff < minSlopeDiff) {
        minSlopeDiff = slopeDiff
        minDiffLines = [ line1, line2 ]
      }
    }
  }

  if (!minDiffLines) {
    minDiffLines = [ lines[0], lines[0] ]
  }

  return lineAverage(minDiffLines[0], minDiffLines[1]) */

  return lineAverage(...lines)
}

/**
 * Get the detection rectangle of an input Mat
 * @param {Mat} inputMat
 * @param {number} detectionRectangleWidth
 * @param {number} detectionRectangleHeight
 * @returns {Mat}
 */
function getDetectionRectangle (inputMat, detectionRectangleWidth, detectionRectangleHeight) {
  let detRectX = (inputMat.cols - detectionRectangleWidth) / 2
  let detRectY = (inputMat.rows - detectionRectangleHeight) / 2

  return inputMat.getRegion(new cv.Rect(
    detRectX,
    detRectY,
    detectionRectangleWidth,
    detectionRectangleHeight
  ))
}

/**
 * Finds the intersection of two lines and returns the point
 * @param {Line} line1
 * @param {Line} line2
 * @return Point2
 */
function findIntersection (line1, line2) {
  let x1 = line1.p1.x
  let x2 = line1.p2.x
  let x3 = line2.p1.x
  let x4 = line2.p2.x

  let y1 = line1.p1.y
  let y2 = line1.p2.y
  let y3 = line2.p1.y
  let y4 = line2.p2.y

  let u = ((x4 - x3) * (y1 - y3) - (y4 - y3) * (x1 - x3)) / ((y4 - y3) * (x2 - x1) - (x4 - x3) * (y2 - y1))

  return new cv.Point2(x1 + u * (x2 - x1), y1 + u * (y2 - y1))
}

/**
 * Creates an average line from multiple lines
 * @param {Line} lines
 * @returns {Line}
 */
function lineAverage (...lines) {
  let sumX1 = 0
  let sumY1 = 0
  let sumX2 = 0
  let sumY2 = 0

  for (let line of lines) {
    sumX1 += line.p1.x
    sumX2 += line.p2.x
    sumY1 += line.p1.y
    sumY2 += line.p2.y
  }

  return new Line(
    new cv.Point2(Math.round(sumX1 / lines.length), Math.round(sumY1 / lines.length)),
    new cv.Point2(Math.round(sumX2 / lines.length), Math.round(sumY2 / lines.length))
  )
}

/**
 * Gets the detection borders of an input Mat
 * @param {Mat} inputMat
 * @param {number} detectionWidth
 * @returns {BorderRegion[]}
 */
function getDetectionBorders (inputMat, detectionWidth) {
  let top = inputMat.getRegion(new cv.Rect(
    detectionWidth,
    0,
    inputMat.cols - detectionWidth,
    detectionWidth
  ))

  let bottom = inputMat.getRegion(new cv.Rect(
    detectionWidth,
    inputMat.rows - detectionWidth,
    inputMat.cols - detectionWidth,
    detectionWidth
  ))

  let left = inputMat.getRegion(new cv.Rect(
    0,
    detectionWidth,
    detectionWidth,
    inputMat.rows - detectionWidth
  ))

  let right = inputMat.getRegion(new cv.Rect(
    inputMat.cols - detectionWidth,
    detectionWidth,
    detectionWidth,
    inputMat.rows - detectionWidth
  ))

  return [
    new BorderRegion(top, BorderRegion.TOP),
    new BorderRegion(left, BorderRegion.LEFT),
    new BorderRegion(bottom, BorderRegion.BOTTOM),
    new BorderRegion(right, BorderRegion.RIGHT)
  ]
}

/**
 * Warps four points on a Mat to form a rectangle
 * @param {Mat} inputMat Input Mat that should be warped
 * @param {{topLeft: Point2, topRight: Point2, bottomRight: Point2, bottomLeft: Point2}} cornerPoints Points on the Mat
 * @param {number} width Width of the returned image
 * @param {number} height Height of the returned image
 * @returns {Promise<Mat>}
 */
async function warp (inputMat, cornerPoints, width, height) {
  let trans = cv.getPerspectiveTransform([
    cornerPoints.topLeft,
    cornerPoints.topRight,
    cornerPoints.bottomRight,
    cornerPoints.bottomLeft
  ], [
    new cv.Point2(0, 0),
    new cv.Point2(width, 0),
    new cv.Point2(width, height),
    new cv.Point2(0, height)
  ])

  return inputMat.warpPerspectiveAsync(trans, new cv.Size(width, height))
}

class Line {
  /**
   * Constructs a new Line instance
   * @param p1 First point of the line
   * @param p2 Second point of the line
   * @param slope Slope of the line. If null, the slope will be calculated.
   */
  constructor (p1, p2, slope = null) {
    this.p1 = p1
    this.p2 = p2

    if (!slope) {
      if (p1.x - p2.x < p1.y - p2.y) { // Horizontal
        this.slope = Math.abs((p1.y - p2.y) / (p1.x - p2.x))
      } else { // Vertical
        this.slope = Math.abs((p1.x - p2.x) / (p1.y - p2.y))
      }
    } else this.slope = slope
  }
}

class BorderRegion {
  static get TOP () { return 1 }
  static get LEFT () { return 2 }
  static get BOTTOM () { return 3 }
  static get RIGHT () { return 4 }

  /**
   * Constructs a new border region to detect lines in
   * @param {Mat} regionMat Mat of the region
   * @param {number} type Either one of BorderRegion::TOP, BorderRegion::LEFT, BorderRegion::BOTTOM or BorderRegion::RIGHT
   */
  constructor (regionMat, type) {
    this.regionMat = regionMat
    this.type = type
  }
}
